### ISSUE TITLE
Getting "ZeroDivisionError: division by zero" in training

### ISSUE THREAD

**Issue by Ayushkumawat** [ShawkhIbneRashid/medical-images-sr#1]
I am getting ZeroDivisionError: division by zero while training the model in - "real_esrgan_notebook.ipynb" file

![Screenshot 2023-01-18 234053](https://user-images.githubusercontent.com/76219349/213260916-7d07ff42-d5d9-4a0a-85d1-82f9a5517c62.png)

Full Output :- 

Disable distributed.
2023-01-18 17:55:56,857 INFO: 
                ____                _       _____  ____
               / __ ) ____ _ _____ (_)_____/ ___/ / __ \
              / __  |/ __ `// ___// // ___/\__ \ / /_/ /
             / /_/ // /_/ /(__  )/ // /__ ___/ // _, _/
            /_____/ \__,_//____//_/ \___//____//_/ |_|
     ______                   __   __                 __      __
    / ____/____   ____   ____/ /  / /   __  __ _____ / /__   / /
   / / __ / __ \ / __ \ / __  /  / /   / / / // ___// //_/  / /
  / /_/ // /_/ // /_/ // /_/ /  / /___/ /_/ // /__ / /<    /_/
  \____/ \____/ \____/ \____/  /_____/\____/ \___//_/|_|  (_)
    
Version Information: 
	BasicSR: 1.4.2
	PyTorch: 1.13.1+cu116
	TorchVision: 0.14.1+cu116
2023-01-18 17:55:56,858 INFO: 
  name: train_RealESRGAN_in_ou_1_fft_new
  model_type: RealESRGANModel
  scale: 4
  num_gpu: 1
  manual_seed: 0
  l1_gt_usm: True
  percep_gt_usm: True
  gan_gt_usm: False
  resize_prob: [0.2, 0.7, 0.1]
  resize_range: [0.15, 1.5]
  gaussian_noise_prob: 0.5
  noise_range: [1, 30]
  poisson_scale_range: [0.05, 3]
  gray_noise_prob: 0.4
  jpeg_range: [30, 95]
  second_blur_prob: 0.8
  resize_prob2: [0.3, 0.4, 0.3]
  resize_range2: [0.3, 1.2]
  gaussian_noise_prob2: 0.5
  noise_range2: [1, 25]
  poisson_scale_range2: [0.05, 2.5]
  gray_noise_prob2: 0.4
  jpeg_range2: [30, 95]
  gt_size: 256
  queue_size: 180
  datasets:[
    train:[
      name: DIV2K
      type: RealESRGANDataset
      dataroot_gt: datasets/DIV2K
      meta_info: datasets/DIV2K/meta_info/meta_info_DF2Kmultiscale.txt
      io_backend:[
        type: disk
      ]
      blur_kernel_size: 21
      kernel_list: ['iso', 'aniso', 'generalized_iso', 'generalized_aniso', 'plateau_iso', 'plateau_aniso']
      kernel_prob: [0.45, 0.25, 0.12, 0.03, 0.12, 0.03]
      sinc_prob: 0.1
      blur_sigma: [0.2, 3]
      betag_range: [0.5, 4]
      betap_range: [1, 2]
      blur_kernel_size2: 21
      kernel_list2: ['iso', 'aniso', 'generalized_iso', 'generalized_aniso', 'plateau_iso', 'plateau_aniso']
      kernel_prob2: [0.45, 0.25, 0.12, 0.03, 0.12, 0.03]
      sinc_prob2: 0.1
      blur_sigma2: [0.2, 1.5]
      betag_range2: [0.5, 4]
      betap_range2: [1, 2]
      final_sinc_prob: 0.8
      gt_size: 256
      use_hflip: True
      use_rot: False
      use_shuffle: True
      num_worker_per_gpu: 2
      batch_size_per_gpu: 1
      dataset_enlarge_ratio: 1
      prefetch_mode: None
      phase: train
      scale: 4
    ]
  ]
  network_g:[
    type: RRDBNet
    num_in_ch: 1
    num_out_ch: 1
    num_feat: 64
    num_block: 23
    num_grow_ch: 32
    scale: 4
  ]
  network_d:[
    type: UNetDiscriminatorSN
    num_in_ch: 1
    num_feat: 64
    skip_connection: True
  ]
  path:[
    param_key_g: params_ema
    strict_load_g: False
    resume_state: None
    experiments_root: /content/drive/MyDrive/Colab Notebooks/ESRGan/Real-ESRGAN/experiments/train_RealESRGAN_in_ou_1_fft_new
    models: /content/drive/MyDrive/Colab Notebooks/ESRGan/Real-ESRGAN/experiments/train_RealESRGAN_in_ou_1_fft_new/models
    training_states: /content/drive/MyDrive/Colab Notebooks/ESRGan/Real-ESRGAN/experiments/train_RealESRGAN_in_ou_1_fft_new/training_states
    log: /content/drive/MyDrive/Colab Notebooks/ESRGan/Real-ESRGAN/experiments/train_RealESRGAN_in_ou_1_fft_new
    visualization: /content/drive/MyDrive/Colab Notebooks/ESRGan/Real-ESRGAN/experiments/train_RealESRGAN_in_ou_1_fft_new/visualization
  ]
  train:[
    ema_decay: 0.999
    optim_g:[
      type: Adam
      lr: 0.0001
      weight_decay: 0
      betas: [0.9, 0.99]
    ]
    optim_d:[
      type: Adam
      lr: 0.0001
      weight_decay: 0
      betas: [0.9, 0.99]
    ]
    scheduler:[
      type: MultiStepLR
      milestones: [400000]
      gamma: 0.5
    ]
    total_iter: 250000
    warmup_iter: -1
    pixel_opt:[
      type: L1Loss
      loss_weight: 1.0
      reduction: mean
    ]
    perceptual_opt:[
      type: PerceptualLoss
      layer_weights:[
        conv1_2: 0.1
        conv2_2: 0.1
        conv3_4: 1
        conv4_4: 1
        conv5_4: 1
      ]
      vgg_type: vgg19
      use_input_norm: True
      perceptual_weight: 1.0
      style_weight: 0
      range_norm: False
      criterion: l1
    ]
    gan_opt:[
      type: GANLoss
      gan_type: vanilla
      real_label_val: 1.0
      fake_label_val: 0.0
      loss_weight: 0.1
    ]
    net_d_iters: 1
    net_d_init_iters: 0
  ]
  logger:[
    print_freq: 100
    save_checkpoint_freq: 5000.0
    use_tb_logger: True
    wandb:[
      project: None
      resume_id: None
    ]
  ]
  dist_params:[
    backend: nccl
    port: 29500
  ]
  dist: False
  rank: 0
  world_size: 1
  auto_resume: True
  is_train: True
  root_path: /content/drive/MyDrive/Colab Notebooks/ESRGan/Real-ESRGAN

2023-01-18 17:55:59,338 INFO: Dataset [RealESRGANDataset] - DIV2K is built.
Traceback (most recent call last):
  File "realesrgan/train.py", line 11, in <module>
    train_pipeline(root_path)
  File "/usr/local/lib/python3.8/dist-packages/basicsr/train.py", line 120, in train_pipeline
    result = create_train_val_dataloader(opt, logger)
  File "/usr/local/lib/python3.8/dist-packages/basicsr/train.py", line 48, in create_train_val_dataloader
    total_epochs = math.ceil(total_iters / (num_iter_per_epoch))
ZeroDivisionError: division by zero

**Comment by ShawkhIbneRashid** [comment]
I think ZeroDivisonError is occurring as num_iter_per_epoch is zero. Make sure you change the settings properly in the train_realesrgan_x4plus.yml file i.e. the location of your dataset.

**Comment by Ayushkumawat** [comment]
I've replaced "realesrgan_dataset.py, realesrgan_model.py, and utils.py" in 'realesrgan' folder and also replaced "train_realesrgan_x4plus.yml" in 'options' folder under 'Real-ESRGAN' by the files you've uploaded. Is there anything else to be replaced? or do I have to edit the information in the file? 

Also, is there any pattern or numbering in running those files? Like I first tried running 'real_esrgan_notebook.ipynb' and replaced those files in the directory, what's the order? and also what the dataset.ipynb file meant for?
